test/integration/remote-config.spec.ts

```diff
@@ -0,0 +1,259 @@
+        .to.throw(`Invalid Remote Config template: null`);
+    });
+
+    it('should propagate API errors', () => {
+      // rejects with invalid-argument when conditions used in parameters do not exist
+      currentTemplate.conditions = [];
+      currentTemplate.parameters = VALID_PARAMETERS;
+      return admin.remoteConfig().validateTemplate(currentTemplate)
+        .should.eventually.be.rejected.and.have.property('code', 'remote-config/invalid-argument');
+    });
+  });
+
+  describe('publishTemplate', () => {
+    it('should succeed with a vaild template', () => {
+      // set parameters and conditions
+      currentTemplate.conditions = VALID_CONDITIONS;
+      currentTemplate.parameters = VALID_PARAMETERS;
+      return admin.remoteConfig().publishTemplate(currentTemplate)
+        .then((template) => {
+          expect(template.etag).matches(/^etag-[0-9]*-[0-9]*$/);
+          expect(template.conditions.length).to.equal(2);
+          expect(template.conditions).to.deep.equal(VALID_CONDITIONS);
+          expect(template.parameters).to.deep.equal(VALID_PARAMETERS);
+        });
+    });
+
+    INVALID_PARAMETERS.forEach((invalidParameter) => {
+      it(`should throw if the parameters is ${JSON.stringify(invalidParameter)}`, () => {
+        (currentTemplate as any).parameters = invalidParameter;
+        currentTemplate.conditions = VALID_CONDITIONS;
+        expect(() => admin.remoteConfig().publishTemplate(currentTemplate))
+          .to.throw('Remote Config parameters must be a non-null object');
+      });
+    });
+
+    INVALID_CONDITIONS.forEach((invalidConditions) => {
+      it(`should throw if the parameters is ${JSON.stringify(invalidConditions)}`, () => {
+        (currentTemplate as any).conditions = invalidConditions;
+        currentTemplate.parameters = VALID_PARAMETERS;
+        expect(() => admin.remoteConfig().publishTemplate(currentTemplate))
+          .to.throw('Remote Config conditions must be an array');
+      });
+    });
+
+    INVALID_TEMPLATES.forEach((invalidTemplate) => {
+      it(`should throw if the template is ${JSON.stringify(invalidTemplate)}`, () => {
+        expect(() => admin.remoteConfig().publishTemplate(invalidTemplate))
+          .to.throw('ETag must be a non-empty string.');
+      });
+    });
+
+    it(`should throw if the template is null`, () => {
+      expect(() => admin.remoteConfig().publishTemplate(null as any))
+        .to.throw(`Invalid Remote Config template: null`);
+    });
+
+    it('should propagate API errors', () => {
+      // rejects with invalid-argument when conditions used in parameters do not exist
+      currentTemplate.conditions = [];
+      currentTemplate.parameters = VALID_PARAMETERS;
+      return admin.remoteConfig().publishTemplate(currentTemplate)
+        .should.eventually.be.rejected.and.have.property('code', 'remote-config/invalid-argument');
+    });
+  });
+
+  describe('getTemplate', () => {
```

Comment:
Can you go through the docs and make sure that Mocha orders these tests the way you're expecting?

Reply:
According to the docs the tests are run in order they are defined.
https://mochajs.org/#run-cycle-overview
```Worth noting that all hooks, `describe` and `it` callbacks are run in the order they are defined (i.e. found in the file).```

We probably could to a `publishTemplate()` in a `before` block in the same `describe` as the `getTemplate()` to be sure. WDYT?

Reply:
If the behavior is documented we can depend on it. Thanks for checking.